Fix: Cb is not defined if createList fails

SeanDunford opened the pull request:
# Background 
If a user uses `createList` or `buildList` with `async/await` but it fails Mongo Validation the user will see 

`cb is not defined` instead of a rejected promise. 

eg: 

`const users = await monky.createList('users', 3, { name: 'blah blah'});` 
If there was a unique constraint on name and the mongo driver raised 

```
MongoError: E11000 duplicate key error collection:
```

We would expect to see the MongoError and not the cb undefined error. 


# What changed

Return the rejected promise instead of executing undefined callback.

behrendtio commented:
Thanks! Would you mind adding a test that tests this exact scenario? And I don't think we need to add all those dependencies to the package?

SeanDunford commented:
@behrendtio  - Yeah, no problem. 

All i did was `npm install` So i think we might need to pin a bunch of deps or update them in a separate pr.

behrendtio commented:
> Yeah, no problem.

Cheers mate

> All i did was npm install So i think we might need to pin a bunch of deps or update them in a separate or.

Ah no, that's because you're likely running npm v7 and with that, lockfile version 2 - `monky` is still on v1 - the deps haven't changed, only the format has. So all good then 👍

SeanDunford commented:
Hey dude I added the regression test to ensure the promise rejects appropriately. 

Unfortunately i couldn't replicate with the uniqueness. Uniqueness is enforced at the index level and the specs are stubbing save from writing to the database  in `test/stub.js`.

https://mongoosejs.com/docs/validation.html#the-unique-option-is-not-a-validator


I was able to replicate with a basic validation error and was able to reproduce with the same test. 

<img width="957" alt="image" src="https://user-images.githubusercontent.com/1704718/117396208-ba6e3e00-aeb6-11eb-99f2-9c71bc0a7e9f.png">


And after the change provided in this pr it returns the correct error. 

<img width="1025" alt="image" src="https://user-images.githubusercontent.com/1704718/117396267-d245c200-aeb6-11eb-84e8-8741fd6a9e2e.png">

Added an assertion so the test passes as expected. 

<img width="776" alt="image" src="https://user-images.githubusercontent.com/1704718/117396819-02419500-aeb8-11eb-8f4f-a5798096a37b.png">

SeanDunford commented:
Also, i reverted the `package-lock.json` change as it seems like the code should be updated to support modern node/npm in a seperate pr.

behrendtio commented:
Nice one, thank you. Released with `0.6.13`.

SeanDunford commented:
@behrendtio  Do you have any plans for updates to `monky`? There's a few quirks i tried to highlight in issues. Would be nice to see the project updates to modern js or ts and have better ts typings for better intellisense.

behrendtio commented:
Tbh I didn't have any plans as I frankly haven't used node.js in maybe 4 years professionally. But also `monky` is rather small, so there weren't many bug reports or anything...

You raised a couple of interesting edge cases that don't work though and yeah, the JS is definitely on the more ancient side of things, so wouldn't hurt to bring it into the 2020s...

Let me see how much time I can invest in the next few days to spice it up a bit.
